File: scripts/quickbooks/report_parser.py
"""Parse QuickBooks Profit & Loss and Balance Sheet report JSON into covenant inputs."""
from __future__ import annotations

from typing import Any


def _parse_amount(raw: str | None) -> float:
    if raw is None:
        return 0.0
    cleaned = str(raw).strip().replace(",", "").replace("$", "")
    if cleaned in ("", "-", "—"):
        return 0.0
    negative = cleaned.startswith("(") and cleaned.endswith(")")
    if negative:
        cleaned = cleaned[1:-1]
    try:
        value = float(cleaned)
    except ValueError:
        return 0.0
    return -value if negative else value


def _walk_rows(node: Any, out: dict[str, float]) -> None:
    if not isinstance(node, dict):
        return

    rows = node.get("Row")
    if rows is None:
        return
    if not isinstance(rows, list):
        rows = [rows]

    for row in rows:
        if not isinstance(row, dict):
            continue

        header = row.get("Header", {})
        if header:
            cols = header.get("ColData", [])
            if cols:
                label = str(cols[0].get("value", "")).strip()
                amount = _parse_amount(cols[-1].get("value") if len(cols) > 1 else None)
                if label:
                    out[label.lower()] = amount

        summary = row.get("Summary", {})
        if summary:
            cols = summary.get("ColData", [])
            if cols:
                label = str(cols[0].get("value", "")).strip()
                amount = _parse_amount(cols[-1].get("value") if len(cols) > 1 else None)
                if label:
                    out[label.lower()] = amount

        col_data = row.get("ColData")
        if col_data and isinstance(col_data, list) and len(col_data) >= 2:
            label = str(col_data[0].get("value", "")).strip()
            amount = _parse_amount(col_data[-1].get("value"))
            if label and label.lower() not in ("", "total"):
                out[label.lower()] = amount

        nested = row.get("Rows")
        if nested:
            _walk_rows(nested, out)


def flatten_report(report: dict[str, Any]) -> dict[str, float]:
    lines: dict[str, float] = {}
    _walk_rows(report.get("Rows", {}), lines)
    return lines


def _find_line(lines: dict[str, float], *patterns: str) -> float:
    for pattern in patterns:
        pat = pattern.lower()
        for label, amount in lines.items():
            if pat in label:
                return amount
    return 0.0


def _sum_matching(lines: dict[str, float], *patterns: str) -> float:
    total = 0.0
    used: set[str] = set()
    for pattern in patterns:
        pat = pattern.lower()
        for label, amount in lines.items():
            if pat in label and label not in used:
                total += amount
                used.add(label)
    return total


def extract_covenant_metrics(
    profit_and_loss: dict[str, Any],
    balance_sheet: dict[str, Any],
    cash_flow: dict[str, Any] | None = None,
) -> dict[str, Any]:
    pl = flatten_report(profit_and_loss)
    bs = flatten_report(balance_sheet)
    cf = flatten_report(cash_flow) if cash_flow else {}

    net_income = _find_line(
        pl,
        "net income",
        "net operating income",
        "net earnings",
    )
    interest_expense = abs(
        _find_line(
            pl,
            "interest expense",
            "interest paid",
            "finance costs",
            "bank charges",
        )
    )
    income_tax = abs(_find_line(pl, "income tax", "tax expense", "taxes"))
    depreciation = abs(_find_line(pl, "depreciation"))
    amortization = abs(_find_line(pl, "amortization"))
    operating_income = _find_line(
        pl,
        "operating income",
        "net operating income",
        "income from operations",
    )
    revenue = _find_line(pl, "total income", "total revenue", "gross profit")
    rent_lease = abs(_find_line(pl, "rent", "lease", "occupancy"))

    ebitda = operating_income + depreciation + amortization
    if ebitda == 0:
        ebitda = net_income + interest_expense + income_tax + depreciation + amortization

    current_assets = _find_line(bs, "total current assets", "current assets")
    current_liabilities = _find_line(
        bs, "total current liabilities", "current liabilities"
    )
    total_assets = _find_line(bs, "total assets")
    total_liabilities = _find_line(bs, "total liabilities")
    total_equity = _find_line(
        bs,
        "total equity",
        "stockholders' equity",
        "shareholders' equity",
        "member's equity",
    )
    if total_equity == 0:
        total_equity = _find_line(bs, "equity")

    cash = _sum_matching(
        bs,
        "cash and cash equivalents",
        "checking",
        "savings",
        "petty cash",
    )
    if cash == 0:
        cash = _find_line(bs, "total bank accounts", "cash")

    long_term_debt = _sum_matching(
        bs,
        "long-term",
        "long term",
        "notes payable",
        "line of credit",
        "term loan",
        "mortgage",
    )
    short_term_debt = _sum_matching(
        bs,
        "short-term",
        "short term",
        "current portion of long-term",
        "credit card",
    )
    total_debt = long_term_debt + short_term_debt
    if total_debt == 0 and total_liabilities > 0:
        total_debt = max(total_liabilities - current_liabilities, 0)

    principal_payments = abs(
        _find_line(cf, "repayment", "principal", "debt service")
        if cf
        else 0.0
    )

    leverage_ratio = round(total_debt / ebitda, 4) if ebitda > 0 else 0.0
    interest_coverage = (
        round(ebitda / interest_expense, 4) if interest_expense > 0 else 0.0
    )
    current_ratio = (
        round(current_assets / current_liabilities, 4)
        if current_liabilities > 0
        else 0.0
    )
    fixed_charge_denominator = interest_expense + rent_lease + principal_payments
    fixed_charge_coverage = (
        round(ebitda / fixed_charge_denominator, 4)
        if fixed_charge_denominator > 0
        else 0.0
    )
    debt_to_equity = (
        round(total_debt / total_equity, 4) if total_equity > 0 else 0.0
    )
    debt_service_coverage = (
        round(
            ebitda / (interest_expense + principal_payments),
            4,
        )
        if (interest_expense + principal_payments) > 0
        else 0.0
    )

    return {
        "source": "quickbooks",
        "raw": {
            "profitAndLossLines": len(pl),
            "balanceSheetLines": len(bs),
            "cashFlowLines": len(cf),
        },
        "reportLines": {
            "profitAndLoss": pl,
            "balanceSheet": bs,
            "cashFlow": cf,
        },
        "figures": {
            "revenue": revenue,
            "netIncome": net_income,
            "operatingIncome": operating_income,
            "interestExpense": interest_expense,
            "incomeTaxExpense": income_tax,
            "depreciation": depreciation,
            "amortization": amortization,
            "ebitda": round(ebitda, 2),
            "currentAssets": current_assets,
            "currentLiabilities": current_liabilities,
            "totalAssets": total_assets,
            "totalLiabilities": total_liabilities,
            "totalEquity": total_equity,
            "cash": cash,
            "longTermDebt": long_term_debt,
            "shortTermDebt": short_term_debt,
            "totalDebt": round(total_debt, 2),
            "rentAndLease": rent_lease,
            "principalPayments": principal_payments,
        },
        "ratios": {
            "leverageRatio": leverage_ratio,
            "interestCoverageRatio": interest_coverage,
            "currentRatio": current_ratio,
            "fixedChargeCoverage": fixed_charge_coverage,
            "debtToEquity": debt_to_equity,
            "debtServiceCoverage": debt_service_coverage,
        },
    }

File: server/quickbooks/reportParser.js
function parseAmount(raw) {
  if (raw == null) return 0;
  let cleaned = String(raw).trim().replace(/,/g, "").replace(/\$/g, "");
  if (cleaned === "" || cleaned === "-" || cleaned === "—") return 0;
  const negative = cleaned.startsWith("(") && cleaned.endsWith(")");
  if (negative) cleaned = cleaned.slice(1, -1);
  const value = Number.parseFloat(cleaned);
  if (Number.isNaN(value)) return 0;
  return negative ? -value : value;
}

function walkRows(node, out) {
  if (!node || typeof node !== "object") return;

  let rows = node.Row;
  if (rows == null) return;
  if (!Array.isArray(rows)) rows = [rows];

  for (const row of rows) {
    if (!row || typeof row !== "object") continue;

    const header = row.Header;
    if (header?.ColData) {
      const label = String(header.ColData[0]?.value ?? "").trim();
      const amount = parseAmount(header.ColData.at(-1)?.value);
      if (label) out[label.toLowerCase()] = amount;
    }

    const summary = row.Summary;
    if (summary?.ColData) {
      const label = String(summary.ColData[0]?.value ?? "").trim();
      const amount = parseAmount(summary.ColData.at(-1)?.value);
      if (label) out[label.toLowerCase()] = amount;
    }

    if (row.ColData?.length >= 2) {
      const label = String(row.ColData[0]?.value ?? "").trim();
      const amount = parseAmount(row.ColData.at(-1)?.value);
      if (label && label.toLowerCase() !== "total") {
        out[label.toLowerCase()] = amount;
      }
    }

    if (row.Rows) walkRows(row.Rows, out);
  }
}

export function flattenReport(report) {
  const lines = {};
  walkRows(report?.Rows ?? {}, lines);
  return lines;
}

function findLine(lines, ...patterns) {
  // First pass: try exact matches
  for (const pattern of patterns) {
    const pat = pattern.toLowerCase();
    for (const [label, amount] of Object.entries(lines)) {
      if (label === pat) return amount;
    }
  }
  
  // Second pass: try prefix matches (avoid matching compound labels like "total liabilities and equity" when searching for "total liabilities")
  for (const pattern of patterns) {
    const pat = pattern.toLowerCase();
    let bestMatch = null;
    let bestLength = -1;
    for (const [label, amount] of Object.entries(lines)) {
      if (label.includes(pat)) {
        // Prefer shorter matches (more specific)
        // For "total liabilities", prefer "total liabilities" over "total liabilities and equity"
        if (bestMatch === null || label.length < bestLength) {
          bestMatch = amount;
          bestLength = label.length;
        }
      }
    }
    if (bestMatch !== null) return bestMatch;
  }
  
  return 0;
}

function sumMatching(lines, ...patterns) {
  let total = 0;
  const used = new Set();
  for (const pattern of patterns) {
    const pat = pattern.toLowerCase();
    for (const [label, amount] of Object.entries(lines)) {
      if (label.includes(pat) && !used.has(label)) {
        total += amount;
        used.add(label);
      }
    }
  }
  return total;
}

export function extractCovenantMetrics(profitAndLoss, balanceSheet, cashFlow = null) {
  const pl = flattenReport(profitAndLoss);
  const bs = flattenReport(balanceSheet);
  const cf = cashFlow ? flattenReport(cashFlow) : {};

  const netIncome = findLine(pl, "net income", "net operating income", "net earnings");
  const interestExpense = Math.abs(
    findLine(pl, "interest expense", "interest paid", "finance costs", "bank charges"),
  );
  const incomeTax = Math.abs(findLine(pl, "income tax", "tax expense", "taxes"));
  const depreciation = Math.abs(findLine(pl, "depreciation"));
  const amortization = Math.abs(findLine(pl, "amortization"));
  const operatingIncome = findLine(
    pl,
    "operating income",
    "net operating income",
    "income from operations",
  );
  const revenue = findLine(pl, "total income", "total revenue", "gross profit");
  const rentLease = Math.abs(findLine(pl, "rent", "lease", "occupancy"));

  let ebitda = operatingIncome + depreciation + amortization;
  if (ebitda === 0) {
    ebitda = netIncome + interestExpense + incomeTax + depreciation + amortization;
  }

  const currentAssets = findLine(bs, "total current assets", "current assets");
  const currentLiabilities = findLine(bs, "total current liabilities", "current liabilities");
  const totalAssets = findLine(bs, "total assets");
  const totalLiabilities = findLine(bs, "total liabilities");
  let totalEquity = findLine(
    bs,
    "total equity",
    "stockholders' equity",
    "shareholders' equity",
    "member's equity",
  );
  if (totalEquity === 0) totalEquity = findLine(bs, "equity");

  // DEBUG: Log balance sheet equation
  const calculatedEquity = totalAssets - totalLiabilities;
  const isBalanced = Math.abs((totalAssets - (totalLiabilities + totalEquity))) < 0.01;
  console.log("[QB Parser Debug] Balance Sheet Check:", {
    totalAssets,
    totalLiabilities,
    totalEquity,
    calculatedEquity,
    isBalanced,
    equation: `${totalAssets} = ${totalLiabilities} + ${totalEquity} = ${totalLiabilities + totalEquity}`,
    allBSLines: bs,
  });

  let cash = sumMatching(bs, "cash and cash equivalents", "checking", "savings", "petty cash");
  if (cash === 0) cash = findLine(bs, "total bank accounts", "cash");

  const longTermDebt = sumMatching(
    bs,
    "long-term",
    "long term",
    "notes payable",
    "line of credit",
    "term loan",
    "mortgage",
  );
  const shortTermDebt = sumMatching(
    bs,
    "short-term",
    "short term",
    "current portion of long-term",
    "credit card",
  );
  let totalDebt = longTermDebt + shortTermDebt;
  if (totalDebt === 0 && totalLiabilities > 0) {
    totalDebt = Math.max(totalLiabilities - currentLiabilities, 0);
  }

  const principalPayments = Math.abs(findLine(cf, "repayment", "principal", "debt service"));

  const leverageRatio = ebitda > 0 ? round(totalDebt / ebitda, 4) : 0;
  const interestCoverageRatio =
    interestExpense > 0 ? round(ebitda / interestExpense, 4) : 0;
  const currentRatio =
    currentLiabilities > 0 ? round(currentAssets / currentLiabilities, 4) : 0;
  const fixedChargeDenominator = interestExpense + rentLease + principalPayments;
  const fixedChargeCoverage =
    fixedChargeDenominator > 0 ? round(ebitda / fixedChargeDenominator, 4) : 0;
  // Debt-to-Equity: if equity is negative or zero, report the magnitude of the negative ratio
  const debtToEquity = totalEquity !== 0 ? round(totalDebt / totalEquity, 4) : (totalDebt > 0 ? 999.9 : 0);
  const debtServiceDenominator = interestExpense + principalPayments;
  const debtServiceCoverage =
    debtServiceDenominator > 0 ? round(ebitda / debtServiceDenominator, 4) : 0;

  return {
    source: "quickbooks",
    raw: {
      profitAndLossLines: Object.keys(pl).length,
      balanceSheetLines: Object.keys(bs).length,
      cashFlowLines: Object.keys(cf).length,
    },
    reportLines: {
      profitAndLoss: pl,
      balanceSheet: bs,
      cashFlow: cf,
    },
    figures: {
      revenue,
      netIncome,
      operatingIncome,
      interestExpense,
      incomeTaxExpense: incomeTax,
      depreciation,
      amortization,
      ebitda: round(ebitda, 2),
      currentAssets,
      currentLiabilities,
      totalAssets,
      totalLiabilities,
      totalEquity,
      cash,
      longTermDebt,
      shortTermDebt,
      totalDebt: round(totalDebt, 2),
      rentAndLease: rentLease,
      principalPayments,
    },
    ratios: {
      leverageRatio,
      interestCoverageRatio,
      currentRatio,
      fixedChargeCoverage,
      debtToEquity,
      debtServiceCoverage,
    },
  };
}

function round(n, digits) {
  const p = 10 ** digits;
  return Math.round(n * p) / p;
}
